eos-updater/tests/eos_updater.py

```diff
@@ -0,0 +1,428 @@
+                dbus.Int64(parameters.get('DownloadedBytes', 0)),
+            'UnpackedSize': dbus.Int64(parameters.get('UnpackedSize', 0)),
+            'FullDownloadSize':
+                dbus.Int64(parameters.get('FullDownloadSize', 0)),
+            'FullUnpackedSize':
+                dbus.Int64(parameters.get('FullUnpackedSize', 0)),
+            'ErrorCode': dbus.UInt32(parameters.get('ErrorCode', 0)),
+            'ErrorName': dbus.String(parameters.get('ErrorName', '')),
+            'ErrorMessage': dbus.String(parameters.get('ErrorMessage', '')),
+        }, signature='sv'))
+
+    # Set up initial state
+    mock.__poll_action = 'no-update'
+    mock.__fetch_action = 'success'
+    mock.__apply_action = 'success'
+
+    # Set up private methods
+    mock.__set_properties = __set_properties
+    mock.__change_state = __change_state
+    mock.__set_error = __set_error
+    mock.__check_state = __check_state
+
+
+#
+# Internal utility methods
+#
+
+# Values in @properties must have variant_level≥1
+def __set_properties(self, iface, properties):
+    for key, value in properties.items():
+        self.props[iface][key] = value
+    self.EmitSignal(dbus.PROPERTIES_IFACE, 'PropertiesChanged', 'sa{sv}as', [
+        iface,
+        properties,
+        [],
+    ])
+
+
+def __change_state(self, new_state):
+    props = {
+        'State': dbus.UInt32(new_state, variant_level=1)
+    }
+
+    # Reset error state if necessary.
+    if new_state != UpdaterState.ERROR and \
+       self.props[MAIN_IFACE]['ErrorName'] != '':
+        props['ErrorCode'] = dbus.UInt32(0, variant_level=1)
+        props['ErrorName'] = dbus.String('', variant_level=1)
+        props['ErrorMessage'] = dbus.String('', variant_level=1)
+
+    self.__set_properties(self, MAIN_IFACE, props)
+    self.EmitSignal(MAIN_IFACE, 'StateChanged', 'u', [dbus.UInt32(new_state)])
+
+
+def __set_error(self, error_name, error_message):
+    assert(error_name != '')
+
+    self.__set_properties(self, MAIN_IFACE, {
+        'ErrorName': dbus.String(error_name, variant_level=1),
+        'ErrorMessage': dbus.String(error_message, variant_level=1),
+        'ErrorCode': dbus.UInt32(1, variant_level=1),
+    })
+    self.__change_state(self, UpdaterState.ERROR)
+
+
+def __check_state(self, allowed_states):
+    if self.props[MAIN_IFACE]['State'] not in allowed_states:
+        raise dbus.exceptions.DBusException(
+            'Call not allowed in this state',
+            name='com.endlessm.Updater.Error.WrongState')
+
+
+#
+# Updater methods which are too big for squeezing into AddMethod()
+#
+
+@dbus.service.method(MAIN_IFACE, in_signature='', out_signature='')
+def Poll(self):
+    self.__check_state(self, set([
+        UpdaterState.READY,
+        UpdaterState.UPDATE_AVAILABLE,
+        UpdaterState.UPDATE_READY,
+        UpdaterState.ERROR,
+    ]))
+
+    self.__change_state(self, UpdaterState.POLLING)
+
+    if self.__poll_action == 'early-error':
+        time.sleep(0.5)
```

Comment:
What's the purpose of the `time.sleep()` calls?

Reply:
To simulate network activity, iirc. I’ll add a comment.